atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/locks/Lockable.java

```diff
@@ -0,0 +1,133 @@
+    // but it should not be true if something local has stopped working on this item and closed the inner class
+    private final AtomicBoolean isLocked = new AtomicBoolean(false);
+
+    private Lockable(
+            T inner, LockDescriptor lockDescriptor, TimelockService timeLock, Refreshable<Duration> lockTimeout) {
+        this.timeLock = timeLock;
+        this.inner = inner;
+        this.lockTimeout = lockTimeout;
+        this.lockDescriptor = lockDescriptor;
+    }
+
+    public static <T> Lockable<T> create(
+            T inner, LockDescriptor lockDescriptor, TimelockService timeLock, Refreshable<Duration> lockTimeout) {
+        return new Lockable<>(inner, lockDescriptor, timeLock, lockTimeout);
+    }
+
+    // For code owners:
+    // Do not use this for multiple locks!! If you're not careful, you can make it really easy to deadlock
+    // If you end up using this class and have a need for batch locks, implement a separate method that uses
+    // Timelocks batch locking methods and ensure there's a consistent ordering when taking out local locks
+    // to avoid circular waiting.
+    //    @MustBeClosed TODO: How do I get this working?? I'll track looking into this for another time.
+    @CheckReturnValue
+    public Optional<Inner> tryLock() {
+        if (isLocked.compareAndSet(false, true)) {
+            // We do not want the timeout to be too low to avoid a race condition where we give up too soon
+            LockRequest request = LockRequest.of(
+                    ImmutableSet.of(lockDescriptor), lockTimeout.get().toMillis());
+
+            try {
+                return timeLock.lock(request).getTokenOrEmpty().map(lockToken -> new Inner(lockToken));
+            } catch (Exception e) {
+                log.error("Failed to acquire lock", UnsafeArg.of("lockDescriptor", lockDescriptor), e);
+                isLocked.set(false);
+                return Optional.empty();
+            }
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    // Expected to be used in a try closeable.
+    public final class Inner implements AutoCloseable {
+        private final LockToken lockToken;
+
+        private Inner(LockToken lockToken) {
+            this.lockToken = lockToken;
+        }
+
+        public T getInner() {
+            return inner;
+        }
+
+        public Lockable<T> getOuter() {
+            return Lockable.this;
+        }
+
+        @Override
+        public void close() {
+            try {
+                timeLock.unlock(ImmutableSet.of(lockToken));
+            } finally {
+                isLocked.set(false); // We end up holding the local lock ever so slightly longer than necessary.
+            }
+        }
+    }
+
+    public static class LockableComparator<T> implements Comparator<Lockable<T>> {
+        private final Comparator<T> comparator;
+
+        public LockableComparator(Comparator<T> comparator) {
```

Comment:
FWIW this used to have a bound T extends Comparable<T> which is why SweepableBuckets extended Comparable. I've changed it now - the sweepable buckets change will come in the PR with the coordinator